lib/SILOptimizer/Utils/Differentiation/Common.cpp

```diff
@@ -23,5 +23,178 @@ namespace autodiff {
 
 raw_ostream &getADDebugStream() { return llvm::dbgs() << "[AD] "; }
 
+//===----------------------------------------------------------------------===//
+// Code emission utilities
+//===----------------------------------------------------------------------===//
+
+void collectAllActualResultsInTypeOrder(
+    ApplyInst *ai, ArrayRef<SILValue> extractedDirectResults,
+    SmallVectorImpl<SILValue> &results) {
+  auto calleeConvs = ai->getSubstCalleeConv();
+  unsigned indResIdx = 0, dirResIdx = 0;
+  for (auto &resInfo : calleeConvs.getResults()) {
+    results.push_back(resInfo.isFormalDirect()
+                          ? extractedDirectResults[dirResIdx++]
+                          : ai->getIndirectSILResults()[indResIdx++]);
+  }
+}
+
+SILValue joinElements(ArrayRef<SILValue> elements, SILBuilder &builder,
+                      SILLocation loc) {
+  if (elements.size() == 1)
+    return elements.front();
+  return builder.createTuple(loc, elements);
+}
+
+void extractAllElements(SILValue value, SILBuilder &builder,
+                        SmallVectorImpl<SILValue> &results) {
+  auto tupleType = value->getType().getAs<TupleType>();
+  if (!tupleType) {
+    results.push_back(value);
+    return;
+  }
+  if (builder.hasOwnership()) {
+    auto *dti = builder.createDestructureTuple(value.getLoc(), value);
+    results.append(dti->getResults().begin(), dti->getResults().end());
+    return;
+  }
+  for (auto i : range(tupleType->getNumElements()))
+    results.push_back(builder.createTupleExtract(value.getLoc(), value, i));
+}
+
+void emitZeroIntoBuffer(SILBuilder &builder, CanType type,
```

Comment:
This again doesn't feel like a method that has anything to do with differentiation.

Reply:
Currently, this code currently emits `AdditiveArithmetic.zero`, but we plan to change it to emit `Differentiable.zeroTangentVector` sometime (TF-1005), so I'd prefer to keep it differentiation-specific.

Callers of this function could be deduped though: `PullbackEmitter::emitZeroIndirect` and `JVPEmitter::emitZeroIndirect`. I can look into that, but I'm not sure how to best use `SILBuilderContext &` since a `SILBasicBlock::iterator` is needed as an insertion point: https://github.com/apple/swift/pull/30818#discussion_r406975815